### DISCUSSION TITLE
Suggestion: null-conditional default value operator

### DISCUSSION THREAD

**Discussion by will-bartlett** [dotnet/csharplang#784] [General]
The `?.` operator is great for working with objects. It's not so great for working with value types.

Consider an certificate that may or may not be null. I would like to check if I have a certificate and if the certificate has a length greater than zero. Today, I write `if (cert != null && cert.Length != 0)`.

Consider an object person that has a boolean property IsEmployee. I would like to know if person is null (anonymous / no authentication and if the person is an employee). Yesterday, I wrote `if (person != null && person.IsEmployee)`. Today, I write `if (person?.IsEmployee == true)`. This works, but is a bit awkward (particularly for `a?.b != true`)

Introducing a new operator `?_`.
`a?_b` means `(a == null : default(typeof(b)) : a.b)`.

So, in my first case, I would write: `if (cert?_Length > 0)`.
In my second case, I would write: `if (person?_IsEmployee)`.

Interestingly, `?_` would actually have the same meaning as `?.` for reference type properties. That is, if `b` is an object then `a?_b` would equal `a?.b` for all possible values of `a`.

**Comment by yaakov-h** [comment]
For your certificate, you can use `cert?.Length > 0`.

For the others... this only seems to take away control from manually specifying the default case, e.g. `person != null && person.IsEmployee` can also be expressed as `person?.IsEmployee ?? false`, where the code explicitly specifies the default when-it's-null case of false.

I personally don't see the value in this.

(I also don't like the use of the underscore here.)

**Comment by will-bartlett** [comment]
compare:
`if (!employeesOnly || (sender?.IsEmployee ?? false) && (recipient?.IsEmployee ?? false))`
to:
`if (!employeesOnly || (sender?_IsEmployee && recipient?_IsEmployee))`
It's not huge, but .NET has default values so that the code doesn't need to explicitly specify the default value.

Didn't know that all nullable type comparisons with null return false. Thanks, that is helpful. I don't particularly care what symbol is used to represent the operator as long as it starts with a question mark.

**Comment by CyrusNajmabadi** [comment]
_IsEmployee is already a valid name in C#.

**Comment by HaloFour** [comment]
.NET has default values, but the result of a null-condition is always a nullable, not a default value.  In most cases it wouldn't make sense for the expression to result in a default value for the case of a value type, and for the cases where it does you already have the option to use the null-coalescing operator (`??`).

**Comment by jnm2** [comment]
Similar to https://github.com/dotnet/csharplang/issues/328. If what's important is to default to a specific value like `false` or `true`, specifying the important value explicitly is best. If for some reason the value isn't important but the fact that it is default is important (can't imagine why), I think you'll be able to express it as `?? default` in C# 7.1.

**Comment by sirgru** [comment]
I think a better solution would be to allow overloading the ?. operator. That way we could specify exactly what we mean by default state, rather than fixing it in the language. IIRC, guys from Unity Engine had this as a major request, they use a "null object" in places, and also override ==, Equals etc. in the base class for all objects. Now since ?. is not overloadable there is an issue with that. I'm pretty sure the C# team knows about this though.
